dao/rhc_connection_dao.go

```diff
@@ -0,0 +1,255 @@
+		Debug().
+		Model(&m.RhcConnection{}).
+		Select(`"rhc_connections".*, STRING_AGG(CAST ("jt"."source_id" AS TEXT), ',') AS "source_ids"`).
+		Joins(`INNER JOIN "source_rhc_connections" AS "jt" ON "rhc_connections"."id" = "jt"."rhc_connection_id"`).
+		Where(`"rhc_connections"."id" = ?`, id).
+		Where(`"jt"."tenant_id" = ?`, s.TenantID).
+		Group(`"rhc_connections"."id"`)
+
+	// Run the actual query.
+	result, err := query.Rows()
+	if err != nil {
+		return nil, err
+	}
+
+	// We call next as otherwise "ScanRows" complains, but since we're going to map the results to an array of
+	// map[string]interface{}, "ScanRows" will already scan every row into that array, thus freeing us from calling
+	// result.Next() again.
+	if !result.Next() {
+		return nil, util.NewErrNotFound("rhcConnection")
+	}
+
+	// Loop through the rows to map both the connection and its related sources.
+	var rows []map[string]interface{}
+	err = DB.ScanRows(result, &rows)
+	if err != nil {
+		return nil, err
+	}
+
+	err = result.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	if len(rows) != 1 {
+		return nil, errors.New("unexpected number of results")
+	}
+
+	rhcConnection, err := mappers.MapRowToRhcConnection(rows[0])
+	if err != nil {
+		return nil, err
+	}
+
+	return rhcConnection, nil
+}
+
+func (s *RhcConnectionDaoImpl) Create(rhcConnection *m.RhcConnection) (*m.RhcConnection, error) {
+	// If the source doesn't exist we cannot create the RhcConnection, since it needs to be linked to at least one
+	// source.
+	var sourceExists bool
+	err := DB.Debug().
+		Model(&m.Source{}).
+		Select(`1`).
+		Where(`id = ?`, rhcConnection.Sources[0].ID).
+		Scan(&sourceExists).
+		Error
+
+	// Something went wrong with the query
+	if err != nil {
+		return nil, err
+	}
+
+	if !sourceExists {
+		return nil, util.NewErrNotFound("source")
+	}
+
+	err = DB.Transaction(func(tx *gorm.DB) error {
+		var err error
+
+		err = tx.Debug().
+			Where(`rhc_id = ?`, rhcConnection.RhcId).
+			Omit(clause.Associations).
+			FirstOrCreate(&rhcConnection).
+			Error
+
+		if err != nil {
+			return err
+		}
+
+		// Try to insert an sourceRhcConnection. If it exists the database will complain.
+		sourceRhcConnection := m.SourceRhcConnection{
+			SourceId:        rhcConnection.Sources[0].ID,
+			RhcConnectionId: rhcConnection.ID,
+			TenantId:        s.TenantID,
+		}
+
+		err = tx.Debug().
+			Create(&sourceRhcConnection).
+			Error
+
+		if err != nil {
+			return fmt.Errorf("cannot link red hat connection to source: %w", err)
+		}
+
+		return nil
+	})
+
+	return rhcConnection, err
+}
+
+func (s *RhcConnectionDaoImpl) Update(rhcConnection *m.RhcConnection) error {
+	err := DB.Debug().
+		Updates(rhcConnection).
+		Error
+	return err
+}
+
+func (s *RhcConnectionDaoImpl) Delete(id *int64) error {
+	rhcConnection := &m.RhcConnection{ID: *id}
+
+	err := DB.Debug().
+		Where("id = ?", id).
+		First(&rhcConnection).
+		Error
+
+	if err != nil {
+		return util.NewErrNotFound("rhcConnection")
+	}
```

Comment:
Hmmm...this got em thinking of a refactor we can do eventually. Rather than fetching the entirety of the record and checking I think we can use (read as: abuse) the postgres RETURNING keyword to do it all in one. 

EDIT: yep found the docs! It's ugly but easy enough to understand: https://gorm.io/docs/delete.html#Returning-Data-From-Deleted-Rows

We can do that in a refactor though, that way we can change it everywhere. It will be nice to change out all of the `Delete` functions across the app to use it.

Reply:
I've refactored the `.Delete()` function so it returns the deleted object, because I had missed it.

---

I also was thinking on creating an issue to implement the `.Exists() (bool, err)` function in the DAOs, which only would do a `SELECT 1` instead of a  `SELECT *`, since there are handlers which only need to know whether the related entity exists or not.

Anyway, I created a story for the refactor so we don't forget: https://issues.redhat.com/browse/RHCLOUD-18138